### DISCUSSION TITLE
Specifying the Direction of a Traction BC applied at Nodes for a Solid Mechanics Linearized-Elasticity Problem in 3D

### DISCUSSION THREAD

**Discussion by kurtenkera** [idaholab/moose#27714] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [x] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

I'm trying to solve the 3D linearized-elasticity PDE seen below:

![image](https://github.com/idaholab/moose/assets/127472392/d0db094c-338e-4804-81be-7fbce5d06036)

Specifically, I am trying to solve it for the two-dimensional 2 x 2 square geometry seen below. I intend to fix $u_x = u_y = 0$ along the left boundary. Furthermore, I know that my geometry experiences contact forces $\mathbf{F_1}$ and $\mathbf{F_2}$ in the directions shown below at the upper right and lower right corners of the square, where $\|\mathbf{F_1}\| = 5000000N$ and $\|\mathbf{F_2}\| = 2000000N$.

![FEA_problem](https://github.com/idaholab/moose/assets/127472392/6d3c40c2-2183-41d7-8494-05e49fe4ed33)

**Question: How do I specify these force/traction boundary conditions (i.e., $\mathbf{\sigma} \cdot \mathbf{n} = \mathbf{t}$) using MOOSE, correctly capturing both the desired direction and magnitude of these contact forces?**

I have attached my attempt below, which is clearly wrong:

```
[GlobalParams]
  displacements = 'disp_x disp_y'
[]

[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 20
    ny = 20
    xmin = 0
    xmax = 2
    ymin = 0
    ymax = 2
  []
  [node1]
    type = ExtraNodesetGenerator
    input = gen
    new_boundary = "node1"
    coord = '2 2 0'
  []
  [node2]
    type = ExtraNodesetGenerator
    input = node1
    new_boundary = "node2"
    coord = '2 0 0'
  []
[]

[Physics/SolidMechanics/QuasiStatic]
  [all]
    add_variables = true
  []
[]

[BCs]
  [simply_support_x]     # setting u_x = 0 along the left boundary
    type = DirichletBC
    variable = disp_x
    boundary = 'left'
    value = 0.0
  []
  [simply_support_y]   # setting u_y = 0 along the left boundary
    type = DirichletBC
    variable = disp_y
    boundary = 'left'
    value = 0.0
  []
  [node1force]
    type = ConstantRate
    variable = disp_y
    boundary = "node1"
    rate = 5000000.0
  []
  [node2force]
    type = ConstantRate
    variable = disp_x
    boundary = "node2"
    rate = 2000000.0
  []
[]

[Materials]
  [elasticity]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 1e9
    poissons_ratio = 0.3
  []
  [stress]
    type = ComputeLinearElasticStress
  []
[]

[Executioner]
  type = Steady
[]

[Outputs]
  exodus = true
[]
```

This is wrong because this is the resulting output I get in Paraview:

![image](https://github.com/idaholab/moose/assets/127472392/c05b37d4-af22-4ca7-a22e-4873d674bf94)

Clearly, the above applies the forces $\|\mathbf{F_1}\| = 5000000N$ and $\|\mathbf{F_2}\| = 2000000N$ at the correct nodes, but the direction of $\|\mathbf{F_1}\|$ is applied along the +y-axis, while the direction of $\|\mathbf{F_2}\|$ is applied along the +x-axis, whereas I need the directions pointing diagonally towards the centre of the box as seen in my diagram earlier?

I need to be able to specify the direction which these contact forces are applied at the corner nodes? I inspected the [BCs system](https://mooseframework.inl.gov/syntax/BCs/#:~:text=The%20BCs%20system%20is%20used,may%20be%20specified%20at%20once.), but I can't find anything that accomplishes what I want (I'm guessing there is a solution, but I can't find it as I'm relatively new to solving 3D linearized-elasticity PDEs, and I am still learning about the stress tensor $\mathbf{\sigma}$). Come to think of it, I don't really understand what something like $\mathbf{\sigma} \cdot \mathbf{n}$ means, given that $\mathbf{\sigma}$ is a 3x3 matrix and $\mathbf{n}$ is a vector (i.e., how do we compute a dot product between a matrix and vector)? Perhaps the answer to this might clarify matters for me?

**Comment by GiudGiud** [comment]
Hello

I'm not sure about using ConstantRate. Where did you see this being used? Which example?

I would expect a PressureBC to work for this
There's a Pressure syntax that is a little different you ll want to look at examples

Guillaume

**Comment by maxnezdyur** [comment]
You will need 2 BCs for each of your forces. For a force F, you need to find the F_x and F_y components. Then apply those as BCs. The `ConstantRate` should work for your problem but you need to mind the signs on the forces that you are applying. All of your x forces will be negative, but the bottom node y force will be positive, while the top will be negative still. I would only use the `ConstantRate`, if you are really just trying to apply the forces on two nodes. It would make more sense that the force is spread out, but not sure what you are testing.

**Reply by kurtenkera** [reply]
Thanks @maxnezdyur. Can you help me make sense of how $\mathbf{u}$, $\boldsymbol{\sigma}$ and $\mathbf{F_1}$ are all related? My understanding was that a traction boundary condition is represented by $\boldsymbol{\sigma} \cdot \mathbf{n} = c \in \mathbb{R}$. In my case, $\mathbf{F_1}$ is a traction boundary condition, where $\mathbf{F_1} = \boldsymbol{\sigma} \cdot \mathbf{n} = 5000000N$.

But if the above is true, `ConstantRate` seems to apply $\frac{d\mathbf{u}}{dt} = c$ (i.e. the time-derivative of the displacement-vector), whereas I want a **spatial derivative**, since $\sigma_x = E\frac{\partial u_x}{\partial x}$, where $E$ is Young's Modulus? Thus, I'm not convinced that the `ConstantRate` approach correctly implements these equations since a time derivative is involved?

NOTE: $\mathbf{F_1}$ and $\mathbf{F_2}$ are meant to represent contact forces acting on my square. Eventually, I want to perform an FEA solve on the 3D cube seen below, which is simulated in a physics simulator. My ultimate goal is to be able to perform a solid mechanics FEA solve on **any** arbitrary geometry by (a) simulating it in a physics simulator and (b) reading in the observed contact forces applied at different points/nodes on my mesh as traction BCs applied in MOOSE (so I don't want to have the force spread out, but rather applied at the points observed in the physics simulator). 

https://github.com/idaholab/moose/assets/127472392/5a2dd0ef-510e-4648-bb83-def489295705

**Reply by maxnezdyur** [reply]
The `ConstantRate` docs are giving an example where this could be used. So for the du/dt=c equation the constant rate just adds the "c" part and if you scrolled further down in the example they show another block adds the time derivative. The `ConstantRate` for your problem should just be adding a force directly for each node. A similar object `CoupledForceNodalKernel` would allow you specify a variable to determine the force, but when you look at the implementation they perform the same thing if the coupled variable is constant throughout the domain.

**Reply by kurtenkera** [reply]
> So for the du/dt=c equation the constant rate just adds the "c" part and if you scrolled further down in the example they show another block adds the time derivative.

I can see how we are adding the "c" part. My issue is that I don't understand how we are adding a force directly for each node using the techniques you have described. For example, suppose $\|\mathbf{F_1}\| = \|\mathbf{F_2}\| = 4N$ in the problem above. Then, using the advice you have given, I would include a block like the following in my input file:

```
[NodalKernels]
  [node1force_xforce]
    type = ConstantRate
    variable = disp_x
    boundary = "node1"
    rate = -2
  []
  [node1force_yforce]
    type = ConstantRate
    variable = disp_y
    boundary = "node1"
    rate = -2
  []
  [node2force_xforce]
    type = ConstantRate
    variable = disp_x
    boundary = "node2"
    rate = -2
  []
  [node2force_yforce]
    type = ConstantRate
    variable = disp_y
    boundary = "node2"
    rate = 2
  []
[]
```

What I am confused about is `disp_x` represents $u_x$ and `disp_y` represents $u_y$, so where is the connection to forces $\mathbf{F_1}$ and $\mathbf{F_2}$ coming from if I am inputting `disp_x` and `disp_y` in these sections?

**Reply by GiudGiud** [reply]
The variable parameter in a kernel is more indicative of which equation you are adding this term to, than of which variable you are modifying

The equation for the displacement is a balance of forces. Adding a term there means adding a force, not setting the value of a displacement

**Reply by kurtenkera** [reply]
Do we need to divide the forces $\bf{F_i}$ by the volume (for a 3D case) of the entire object being modelled? Or in this case, since the forces are only acting locally on the individual voxels at the corners, do I divide these forces by the volume of the individual voxels? My understanding is that the body forces $\bf{b}$ must have units of **force/volume** so they must be divided by some kind of volume - but I'm unsure which kind in this case?

**Reply by GiudGiud** [reply]
if you are applying forces on a point then no need to normalize.
If you are applying on a side, then normalize by the side area.
If you are applying inside an element (volumetric term), then normalize by the volume of the element to form a volumetric density of force.

The idea is simply to conserve the integral of the force. 
I think by voxel you mean a node, so no normalization needed afaik
